fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DataApiRequestImpl.java

```diff
@@ -1348,19 +1349,57 @@ protected void validateTopN(String topNRequest,  int topN, LinkedHashSet<OrderBy
 
         if (showFields.size() == 1 && showFields.contains(DimensionFieldSpecifierKeywords.ALL.toString())) {
             // Show all fields
-            return dimension.getDimensionFields();
+            return immutableLinkedHashSet(dimension.getDimensionFields());
         } else if (showFields.size() == 1 && showFields.contains(DimensionFieldSpecifierKeywords.NONE.toString())) {
             // Show no fields
-            return new LinkedHashSet<>();
+            return immutableLinkedHashSet(new LinkedHashSet<>());
         } else if (!showFields.isEmpty()) {
             // Show the requested fields
-            return bindDimensionFields(dimension, showFields);
+            return immutableLinkedHashSet(bindDimensionFields(dimension, showFields));
         } else {
             // Show the default fields
-            return dimension.getDefaultDimensionFields();
+            return immutableLinkedHashSet(dimension.getDefaultDimensionFields());
```

Comment:
@michael-mclawhorn the error appears when the stream toMap corrupts the binding list.